qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -248,54 +248,53 @@ def run(self, dag):
         replace_start_time = time.time()
 
         def apply_translation(dag, wire_map):
-            dag_updated = False
-            for node in dag.op_nodes():
+            out_dag = dag.copy_empty_like()
+            for node in dag.topological_op_nodes():
                 node_qargs = tuple(wire_map[bit] for bit in node.qargs)
                 qubit_set = frozenset(node_qargs)
                 if node.name in target_basis or len(node.qargs) < self._min_qubits:
                     if isinstance(node.op, ControlFlowOp):
                         flow_blocks = []
                         for block in node.op.blocks:
                             dag_block = circuit_to_dag(block)
-                            dag_updated = apply_translation(
+                            updated_dag = apply_translation(
                                 dag_block,
                                 {
                                     inner: wire_map[outer]
                                     for inner, outer in zip(block.qubits, node.qargs)
                                 },
                             )
-                            if dag_updated:
-                                flow_circ_block = dag_to_circuit(dag_block)
```

Comment:
I think that this might be the origin of the test failure. In the original code, `flow_circ_block = dag_to_circuit(dag_block)` but in the updated one `flow_circ_block = dag_to_circuit(updated_dag)`. Replacing `updated_dag` to `dag_block` fixes the test locally for me, however, I think that the correct line should be `flow_circ_block = dag_to_circuit(updated_dag)` (else, why would you update it and not use it?). Do you think this could have been an oversight in the original code and the test just got dragged with it?

Reply:
Hmm, in the original version of this code `dag_updated` was a boolean that indicated whether the input dag to `apply_translation()` was transformed at all. If it was we would add the modified dag to the `flow_blocks` list which then would get updated. Otherwise we would use the original input block unmodified. I guess maybe we want to have some logic in place to replicate the no-substitution path that I kind of clobbered here.